File: scala/leet_code/2958.scala
import scala.collection.mutable

object Solution {
  def maxSubarrayLength(nums: Array[Int], k: Int): Int = {
    var left       = 0
    val counter    = mutable.HashMap[Int, Int]()
    var max_length = 0
    for (right <- nums.indices) {
      counter.put(nums(right), counter.getOrElse(nums(right), 0) + 1)
      while (counter(nums(right)) > k) {
        counter.put(nums(left), counter(nums(left)) - 1)
        if (counter(nums(left)) == 0) {
          counter.remove(nums(left))
        }
        left += 1
      }
      max_length = Math.max(max_length, right - left + 1)
    }
    max_length
  }
}

File: cpp/leet_code/2958.cpp
// Optimizing the code for speed and unrolling loops for efficiency.
#pragma GCC optimize("O3,unroll-loops")

#include <unordered_map>
#include <algorithm>
#include <vector>


class Solution {
public:
    int maxSubarrayLength(std::vector<int> &nums, int k) {
        int left = 0;
        std::unordered_map<int, int> counter;
        int max_length = 0;
        for (int right = 0; right < nums.size(); ++right) {
            ++counter[nums[right]];
            while (counter[nums[right]] > k) {
                --counter[nums[left]];
                if (counter[nums[left]] == 0) {
                    counter.erase(nums[left]);
                }
                ++left;
            }
            max_length = std::max(max_length, right - left + 1);
        }
        return max_length;

    }
};
